duckduckgo/BrowserServicesKit
# Sync form factor specific favorites

Please review the release process for BrowserServicesKit [here](https://app.asana.com/0/1200194497630846/1200837094583426).

**Required**:

Task/Issue URL: https://app.asana.com/0/1201493110486074/1204926049616866/f
iOS PR: https://github.com/duckduckgo/iOS/pull/2029
macOS PR: https://github.com/duckduckgo/macos-browser/pull/1656
What kind of version bump will this require?: Major

**Optional**:

Tech Design URL: https://app.asana.com/0/481882893211075/1205418608802079/f

**Description**:
This change introduces form factor specific favorites to be used when Sync is enabled.
1 favorites folder is replaced by 3 folders – mobile, desktop and unified.
Users always see only 1 folder and it's their form-factor-specific one. Only users with Sync enabled
get to choose whether they want to see the form-factor-specific favorites or unified (desktop + mobile)
favorites. The setting itself is synced between devices in the Sync account.
BookmarkEntity's isFavorite is replaced with isFavorite(on:) taking folder as a parameter.
API for adding to favorites and removing from favorites was updated to take multiple favorites
folders as needed.
FavoritesDisplayMode enum is introduced to manage display mode in clients.
Bookmarks and Favorites related view models used on iOS are updated to take favorites display mode
and reload their data as the mode changes.
In SyncDataProviders, an abstract SettingSyncHandler class was added to support adding
an arbitrary setting (key-value pair) to Sync. It's subclassed by EmailProtectionSyncHandler
and FavoritesDisplayModeSyncHandlerBase (that needs to be subclassed in client apps due to differences
in User Defaults storage between iOS and macOS apps).



**Steps to test this PR**:
These are common steps for iOS and macOS:
1. Launch the app from develop, populate bookmarks and favorites. (Don't enable Sync).
2. Launch the app from this branch. Identify as internal user and enable Sync.
3. Repeat the above on another platform (iOS/macOS) adding your device to the same Sync account.
4. Verify that both apps show different sets of favorites.
5. Go to Sync settings and enable Unified favorites; verify that both apps show the same set of favorites, which is a logical sum of both platforms' favorites.
6. Verify that reordering favorites gets synced properly.
7. Switch back to form-factor-specific favorites and verify that favorites are back to what you've seen in 4.
8. Continue testing the feature according to [remaining product requirements](https://app.asana.com/0/0/1205419025657593/f).




**OS Testing**:


---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
